Add draft of the credential endpoint (#31)
mattrglobal/oidc-client-bound-assertions-spec: spec.md

## Response Types

It is RECOMMENDED that a Credential Request flow use the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowSteps) as defined in OpenID Connect core.

For instances where [implicit flow](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth) is used, the `response_type` of `credential` SHOULD be used.

## Token Endpoint Response

Successful and Error Authentication Response are in the same manor as [OpenID Connect Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html) with the `code` parameter always being returned with the Authorization Code Flow.

On Request to the Token Endpoint the `grant_type` value MUST be `authorization_code` inline with the Authorization Code Flow and the `code` value included as a parameter.

The following is a non-normative example of a response from the token endpoint, whereby the `access_token` authorizes the Holder to request a `credential` from the credential endpoint.

```
{
  "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCI6Ikp..sHQ",
  "token_type": "bearer",
  "expires_in": 86400,
  "id_token": "eyJodHRwOi8vbWF0dHIvdGVuYW50L..3Mz"
}
```
# Credential Endpoint

> The credential endpoint is used by a Holder to request the issuance of a verifiable credential from a Credential Provder (OP).

Reviewer: ```suggestion
The credential endpoint is used by a Holder to request the issuance of a credential from a Credential Provider (OP).
```

The term credential has been defined in this spec to date but not the term `verifiable credential, im not opposed to defining it just think at this current stage `credential` is more accurate

Reviewer: On second thoughts maybe we should restructure this paragraph to be more inline with section [5.3 of OpenID Connect core](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo)

```suggestion
The Credential Endpoint is an OAuth 2.0 Protected Resource that when called, returns Claims about the authenticated End-User in the form of a credential. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication whereby the the `openid_credential` scope was granted.

Communication with the Credential Endpoint MUST utilize TLS. See Section 16.17 for more information on using TLS.

The Credential Endpoint MUST support the use of HTTP POST methods defined in RFC 2616 [RFC2616].

It is recommended that the Credential Endpoint SHOULD enforce presentation of the OAuth2.0 Access Token to be sender constrained [DPOP](https://tools.ietf.org/html/draft-fett-oauth-dpop-04). However the Credential Endpoint MUST also accept Access Tokens as OAuth 2.0 Bearer Token Usage [RFC6750]. (Note: do we remove?)

The Credential Endpoint SHOULD support the use of Cross Origin Resource Sharing (CORS) [CORS] and or other methods as appropriate to enable Java Script Clients to access the endpoint.
```

> The Holder MUST use the `access_token` previously acquired in the token endpoint response from the Credential Provider as an authorization header.  It is recommended that the `access_token` be sender constrained (Note: do we require this or leave as optional based on the Credential Provider's preference, to provide a DPoP example).

Reviewer: ```suggestion
```

## Credential Endpoint Request Parameters

The Holder may provide a signed request object containing the `sub` to be used as the subject for the resulting credential.  When a `sub` claim is present within the request object an associated `sub_jwk` claim MUST also be present of which the request object MUST be signed with, therefore proving control over the `sub`.

> The Holder may also specify the `credential_format` they wish the credential to be returned in, given it is supported by the Credential Provider. If not specified or not supported the Credential Provider will use their preferred default format.

Reviewer: ```suggestion
The Holder may also specify the `credential_format` they wish the credential to be returned in, given it is supported by the Credential Provider. If not specified or not supported the Credential Provider will use their preferred default format. (Note if we are going to have a default we need to specify it)
```

When a signed request is not provided the Credential Provider will use the `sub` associated with the initial `auth` request, where possible.  If a `sub` value is not available an error should be returned by the Credential Provider.

request
: OPTIONAL. A valid OIDC signed JWT request object. The request object is used to provide a `sub` the Holder wishes to be used as the subject of the resulting credential as well as provide proof of control of that `sub`.

A non-normative example of a Signed Credential request.

```
POST /credential HTTP/1.1
Host: https://issuer.example.com
Authorization: Bearer <access-token>

>   request=<signed-jwt-request-obj>

Reviewer: Clarifying what is the mime type of the request body here? JSON because this looks a bit like a query parameter?

Author: Good catch, it does indeed look like a query param but was intending this to be in the body as `application/json`, update incoming :)

## Response Types

It is RECOMMENDED that a Credential Request flow use the [authorization code flow](https://openid.net/specs/openid-connect-core-1_0.html#CodeFlowSteps) as defined in OpenID Connect core.

For instances where [implicit flow](https://openid.net/specs/openid-connect-core-1_0.html#ImplicitFlowAuth) is used, the `response_type` of `credential` SHOULD be used.

## Token Endpoint Response

Successful and Error Authentication Response are in the same manor as [OpenID Connect Core 1.0](https://openid.net/specs/openid-connect-core-1_0.html) with the `code` parameter always being returned with the Authorization Code Flow.

On Request to the Token Endpoint the `grant_type` value MUST be `authorization_code` inline with the Authorization Code Flow and the `code` value included as a parameter.

The following is a non-normative example of a response from the token endpoint, whereby the `access_token` authorizes the Holder to request a `credential` from the credential endpoint.

```
{
  "access_token": "eyJhbGciOiJSUzI1NiIsInR5cCI6Ikp..sHQ",
  "token_type": "bearer",
  "expires_in": 86400,
  "id_token": "eyJodHRwOi8vbWF0dHIvdGVuYW50L..3Mz"
}
```
# Credential Endpoint

> The Credential Endpoint is an OAuth 2.0 Protected Resource that when called, returns Claims about the authenticated End-User in the form of a credential. To obtain the requested Claims about the End-User, the Client makes a request to the UserInfo Endpoint using an Access Token obtained through OpenID Connect Authentication whereby the the `openid_credential` scope was granted.

Reviewer: ```suggestion
The Credential Endpoint is an OAuth 2.0 Protected Resource that when called, returns Claims about the authenticated End-User in the form of a credential. To obtain a credential on behalf of the End-User, the Client makes a request to the Credential Endpoint using an Access Token obtained through OpenID Connect Authentication whereby the the `openid_credential` scope was granted.
```

Communication with the Credential Endpoint MUST utilize TLS. See Section 16.17 for more information on using TLS.

The Credential Endpoint MUST support the use of HTTP POST methods defined in RFC 2616 [RFC2616].

It is recommended that the Credential Endpoint SHOULD enforce presentation of the OAuth2.0 Access Token to be sender constrained [DPOP](https://tools.ietf.org/html/draft-fett-oauth-dpop-04). However the Credential Endpoint MUST also accept Access Tokens as OAuth 2.0 Bearer Token Usage [RFC6750]. (Note: do we remove?)

The Credential Endpoint SHOULD support the use of Cross Origin Resource Sharing (CORS) [CORS] and or other methods as appropriate to enable Java Script Clients to access the endpoint.

## Credential Endpoint Request Parameters

The Holder may provide a signed request object containing the `sub` to be used as the subject for the resulting credential.  When a `sub` claim is present within the request object an associated `sub_jwk` claim MUST also be present of which the request object MUST be signed with, therefore proving control over the `sub`.

> The Holder may also specify the `credential_format` they wish the credential to be returned in, given it is supported by the Credential Provider. If not specified or not supported the Credential Provider will use their preferred default format. (Note if we are going to have a default we need to specify it)

Reviewer: ```suggestion
The Holder may also specify the `credential_format` they wish the returned credential to be formatted as. If the provider receiving the request does not support the requested credential format they it MUST return an error response, as per [xxxxx]. If the credential_format is not specified in the request SHOULD respond with their preferred or default format. (Note if we are going to have a default we need to specify it or is it at the discretion of the provider to determine this?)
```

> When a signed request is not provided the Credential Provider will use the `sub` associated with the initial `auth` request, where possible.  If a `sub` value is not available an error should be returned by the Credential Provider.

Reviewer: ```suggestion
When a signed request is not provided the Credential Provider will use the `sub` associated with the initial `auth` request, where possible.  If a `sub` value is not available the provider MUST return an error response, as per [TODO].
```
